### PR TITLE
Update Copy-Item to copy directory attributes

### PR DESCRIPTION
**Pull request by Joe-Zer0**
# PR Summary
This change updates Copy-Item to also copy Directory attributes.  There are changes to 3 distinct scenarios.  Copying a folder locally, copying a folder to a remote session, and copying a folder from a remote session.

## PR Context
https://github.com/PowerShell/PowerShell/issues/13552

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by iSazonov** [comment]
@Joe-Zer0 Please add in Summary all that is changed in the PR - this helps reviewers.
Also please add a test/tests for local scenario.

**Comment by Joe-Zer0** [comment]
I added a unit test.  But I made a mess of the commit's.  I am new to rebaseing.  I was attempting to add this change to the existing commit.  I believe that is the desired action for this repository.  I would appreciate a point in the right direction to clean them up.  Thanks!

**Comment by iSazonov** [comment]
> I would appreciate a point in the right direction to clean them up. Thanks!

You have no need to rebase until maintainers ask you. You should add new commits.

To clean you should revert the rebase. If you can not do this you can close the PR and restart with new branch and PR.

**Comment by Joe-Zer0** [comment]
I have fix the git commit's to show 2 commit's:  the original, and the addition of the unit test.  I have also updated the summary to mention the 3 different scenarios that were updated by the changes.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Copy-Item.Tests.ps1 [hunk 1]
```diff
@@ -274,6 +279,15 @@ Describe "Validate Copy-Item Remotely" -Tags "CI" {
                 $copiedFilePath | Should -Exist
             }
         }
+
+        It "Copy-Item Locally and Retain Attributes" {
+            $srcDir = GetSourceFolderPath
+            $destDir = GetDestinationFolderPath
+            $hiddenFolder = New-Item -Path (Join-Path -Path $srcDir -ChildPath "HiddenFolder") -ItemType Directory
+            $hiddenFolder.Attributes += "Hidden"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Copy-Item.Tests.ps1:287]: Hidden attribute is not cross-platform. I suggest to use another attribute for the test.

Comment (Joe-Zer0) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Copy-Item.Tests.ps1]: I have updated the test to use the "System" attribute.

**Comment by Joe-Zer0** [comment]
Please let me know if there is anything else I need to be doing.  Thanks!

**Comment by Joe-Zer0** [comment]
I have made the requested changes.  I was able to add a test for copying TO and FROM a remote session.  To help me learn, can you explain why you wanted one parameter as a named parameter, but not the other one?

**Comment by iSazonov** [comment]
> To help me learn, can you explain why you wanted one parameter as a named parameter, but not the other one?

Common practice in C# use self-descriptive names everywhere. If you use a constant like `123`  a coder reader doesn't understand what is `123`. It is better to define a named const and use it:
```c#
    const int DefaultDelay = 123;
    stopTime = startTime + DefaultDelay;
```
If we say about constants in arguments it is better to use names too as named parameters:
```c#
    Create(Capacity: 100);
```
So we get more readable code.

**Comment by Joe-Zer0** [comment]
Please let me know what else needs to be done for this PR.  Thanks!

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Added some comments, but my concern is if this new behavior might be a breaking change.  Are there scenarios where setting directory attributes can fail?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 2]
```diff
@@ -3730,6 +3732,7 @@ private void CopyItemFromRemoteSession(string path, string destinationPath, bool
                                     destinationPath,
                                     force,
                                     recurse,
+                                    ItemInfo,
```
[the remark this answers is not in the document]
Comment (JamesWTruher) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:3735]: yah, i'm also concerned that we're converting the attributes to strings and then back to attributes

**Comment by Joe-Zer0** [comment]
When copying an item from a session, I am no longer converting FileAttributes to a string.  I was unable to pass FileAttributes through in the hashtable, I kept receiving "Copy-Item: Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IO.FileAttributes'.".  I was able to pass it through as byte however.  This removes the string manipulation that was the concern for several of the comments.  I also updated the parameters for CopyDirectoryFromRemoteSession to pass through the Hashtable.  Please let me know what other changes need to be made.  Thanks!

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 3]
```diff
@@ -9210,6 +9224,7 @@ function PSCreateDirectoryOnRemoteSession
                     if ($force)
                     {{
                         Microsoft.PowerShell.Management\New-Item $createDirectoryPath -ItemType Directory -Force | Out-Null
+                        (Get-Item $createDirectoryPath -Force).Attributes = $attributes
```
[the remark this answers is not in the document]
Comment (Joe-Zer0) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:9222]: I believe the intent of the Force flag is to complete the copy operation regardless of initial state. So the new attributes should be applied.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 4]
```diff
@@ -3708,10 +3708,10 @@ private void CopyItemFromRemoteSession(string path, string destinationPath, bool
...
                         foreach (PSObject item in itemsList)
                         {
-                            Hashtable ItemInfo = (Hashtable)item.BaseObject;
-                            string itemName = (string)ItemInfo["Name"];
-                            string itemFullName = (string)ItemInfo["FullName"];
-                            bool isContainer = (bool)ItemInfo["IsDirectory"];
+                            Hashtable itemInfo = (Hashtable)item.BaseObject;
+                            string itemName = (string)itemInfo["Name"];
+                            string itemFullName = (string)itemInfo["FullName"];
+                            bool isContainer = (bool)itemInfo["IsDirectory"];
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:3714]:
```suggestion
                            var itemInfo = (Hashtable)item.BaseObject;
                            var itemName = (string)itemInfo["Name"];
                            var itemFullName = (string)itemInfo["FullName"];
                            var isContainer = (bool)itemInfo["IsDirectory"];
```
Below too.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 5]
```diff
@@ -4065,9 +4070,10 @@ private void CopyFileInfoItem(FileInfo file, string destinationPath, bool force,
 
             if (ShouldProcess(resource, action))
             {
-                // Create destinationPath directory. This will fail if the directory already exists
-                // and Force is not selected.
-                CreateDirectory(destination, false);
+                // Create destinationPath directory and apply attributes of source directory.
+                // This will fail if the directory already exists and Force is not selected.
+                var destDir = CreateDirectory(destination, streamOutput: true);
+                destDir.Attributes = (FileAttributes)sourceDirectoryAttributes;
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:4076]: I wonder if we can cast to FileAttributes here wait not in definition?

Comment (Joe-Zer0) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:4076]: I don't understand the question.

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:4076]: You cast to byte above and then to FileAttributes here - we could cast to FileAttributes right away.

Comment (Joe-Zer0) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:4076]: If I try to cast directly to FileAttributes
`var sourceDirectoryAttributes = (FileAttributes)sourceDirectoryInfo["Attributes"];`
I receive
`Copy-Item: Unable to cast object of type 'System.Byte' to type 'System.IO.FileAttributes'.`

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:4076]: I think we can:
`var sourceDirectoryAttributes = (FileAttributes)(int)sourceDirectoryInfo["Attributes"];`

Here "int" is default underlying type for Enum types.

**Review by PaulHigin — APPROVED** [review]
Overall LGTM.  I am still not sure how setting directory attributes can fail, and if it does whether that should be a terminating or non-terminating error.  But since the committee is Ok with this going in to next preview version as 'experimental', I think it is Ok as is and we can change if needed.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 5 above)
Comment (Joe-Zer0) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: This is now updated to use int instead of byte.

**Comment by iSazonov** [comment]
@Joe-Zer0 Please make the feature experimental. You can find examples by "ExperimentalFeature" keyword in the repo.
In short:
- add new feature in `static ExperimentalFeature()`
- wrap new code with if (ExperimentalFeature.IsEnable("feature name"))
- skip new tests if the new feature is disabled.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs]: @Joe-Zer0 Please use 
`var sourceDirectoryAttributes = (FileAttributes)(int)sourceDirectoryInfo["Attributes"];`

File: test/powershell/Modules/Microsoft.PowerShell.Management/Copy-Item.Tests.ps1 [hunk 6]
```diff
@@ -279,14 +279,15 @@ Describe "Validate Copy-Item Remotely" -Tags "CI" {
...
-
-        It "Copy-Item Locally and Retain Attributes" {
-            $srcDir = GetSourceFolderPath
-            $destDir = GetDestinationFolderPath
-            $testFolder = New-Item -Path (Join-Path -Path $srcDir -ChildPath "TestFolder") -ItemType Directory
-            $testFolder.Attributes += "System"
-            Copy-Item -Path $testFolder -Destination $destDir
-            ValidateCopyItemOperation -filePath $testFolder
+        if (ExperimentalFeature.IsEnabled("PSCopyItemDirectoryAttributes")) {
+            It "Copy-Item Locally and Retain Attributes" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Copy-Item.Tests.ps1:283]: Please use -Skip parameter in `It` blocks

**Review by iSazonov — APPROVED** [review]

**Comment by Joe-Zer0** [comment]
What are next steps?

**Comment by iSazonov** [comment]
> What are next steps?

Waiting @daxian-dbw ...

**Comment by Joe-Zer0** [comment]
Is there any update on this?

**Comment by iSazonov** [comment]
I hope MSFT team will review community PRs after PowerShell 7.1 release in weeks.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 7]
```diff
@@ -9649,11 +9670,12 @@ function PSGetPathItems
                     return $op
                 }}
 
-                $items = @(Microsoft.PowerShell.Management\Get-Item -Path $getPathItems | Microsoft.PowerShell.Core\ForEach-Object {{
+                $items = @(Microsoft.PowerShell.Management\Get-Item -Path $getPathItems -Force | Microsoft.PowerShell.Core\ForEach-Object {{
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:9673]: Why is `-Force` added?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 8]
```diff
@@ -9649,11 +9670,12 @@ function PSGetPathItems
...
                     @{{
                         FullName = ConvertToPSDrivePath $_.PSDrive $_.FullName;
                         Name = $_.Name;
-                        FileSize = $_.Length; IsDirectory = $_ -is [System.IO.DirectoryInfo]
+                        FileSize = $_.Length; IsDirectory = $_ -is [System.IO.DirectoryInfo];
+                        Attributes = [int]$_.Attributes;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:9678]:
```suggestion
                        FileSize = $_.Length;
                        IsDirectory = $_ -is [System.IO.DirectoryInfo];
                        Attributes = [int]$_.Attributes;
```

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 9]
```diff
@@ -9694,23 +9716,25 @@ function PSGetPathDirAndFiles
 
             try
             {{
-                $item = Microsoft.PowerShell.Management\Get-Item $getPathDir
+                $item = Microsoft.PowerShell.Management\Get-Item $getPathDir -Force
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:9719]: Why is `-Force` added?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 10]
```diff
@@ -9694,23 +9716,25 @@ function PSGetPathDirAndFiles
...
 
                 if ($item -isnot [System.IO.DirectoryInfo])
                 {{
                     return $op
                 }}
 
-                $files = @(Microsoft.PowerShell.Management\Get-ChildItem -Path $getPathDir -File | Microsoft.PowerShell.Core\ForEach-Object {{
+                $files = @(Microsoft.PowerShell.Management\Get-ChildItem -Path $getPathDir -File -Force | Microsoft.PowerShell.Core\ForEach-Object {{
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:9726]: same here, why is `-Force` added?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 11]
```diff
@@ -9694,23 +9716,25 @@ function PSGetPathDirAndFiles
...
                     @{{ FileName = $_.Name;
                         FilePath = (ConvertToPSDrivePath $_.PSDrive $_.FullName);
-                        FileSize = $_.Length
+                        FileSize = $_.Length;
+                        Attributes = [int]$_.Attributes;
                      }}
                 }})
 
-                $directories = @(Microsoft.PowerShell.Management\Get-ChildItem -Path $getPathDir -Directory | Microsoft.PowerShell.Core\ForEach-Object {{
+                $directories = @(Microsoft.PowerShell.Management\Get-ChildItem -Path $getPathDir -Directory -Force | Microsoft.PowerShell.Core\ForEach-Object {{
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:9734]: ditto, why is `-Force` added?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 12]
```diff
@@ -9210,6 +9223,10 @@ function PSCreateDirectoryOnRemoteSession
                     if ($force)
                     {{
                         Microsoft.PowerShell.Management\New-Item $createDirectoryPath -ItemType Directory -Force | Out-Null
+                        if (ExperimentalFeature.IsEnabled('PSCopyItemDirectoryAttributes'))
+                        {{
+                            (Get-Item $createDirectoryPath -Force).Attributes = $attributes
+                        }}
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:9229]: When the remote session is an older version of PowerShell, `ExperimentalFeature.IsEnabled(PSCopyItemDirectoryAttributes)` won't work. I think you will have to decide whether to set the attribute depending on if `$attribute` is `null` or `not-null`. Currently `$attributes` is of a enum type, so I guess you cannot just pass in null when the feature is disabled.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 13]
```diff
@@ -9220,6 +9237,10 @@ function PSCreateDirectoryOnRemoteSession
                 else
                 {{
                     Microsoft.PowerShell.Management\New-Item $createDirectoryPath -ItemType Directory | Out-Null
+                    if (ExperimentalFeature.IsEnabled('PSCopyItemDirectoryAttributes'))
+                    {{
+                        (Get-Item $createDirectoryPath -Force).Attributes = $attributes
+                    }}
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:9243]: Same here, when the remote side is an older version of PS, `ExperimentalFeature.IsEnabled('PSCopyItemDirectoryAttributes')` won't work.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 14]
```diff
@@ -4856,10 +4863,11 @@ private bool RemoteDestinationPathIsFile(string destination, System.Management.A
             return (bool)(op["IsFileInfo"]);
         }
 
-        private string CreateDirectoryOnRemoteSession(string destination, bool force, System.Management.Automation.PowerShell ps)
+        private string CreateDirectoryOnRemoteSession(DirectoryInfo sourceDirectory, string destination, bool force, System.Management.Automation.PowerShell ps)
         {
             ps.AddCommand(CopyFileRemoteUtils.PSCopyToSessionHelperName);
             ps.AddParameter("createDirectoryPath", destination);
+            ps.AddParameter("attributes", sourceDirectory.Attributes);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:4870]: To put this feature under experimental, you will have to guard this line in a `if (ExperimentalFeature.IsEnabled(....))`.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 15]
```diff
@@ -8886,6 +8894,10 @@ internal static class CopyFileRemoteUtils
 
             [string] $createDirectoryPath,
 
+            [Parameter(ParameterSetName=""PSCreateDirectoryOnRemoteSession"", Mandatory=$true)]
```
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8897]: To make the feature experimental, you will need to decide whether to pass in `-Attributes` based on if the experimental feature is turned on, so this cannot be `Mandatory`.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
The remote side could be an older version of PowerShell, so checking `ExperimentalFeature.IsEnabled(...)` on the remote side won't work.

**Comment by daxian-dbw** [comment]
@iSazonov Please don't open this PR. It's been marked as "Waiting on Author" for 25 days, and the author didn't get to it. The author can re-open when ready to work on it again.

**Pull request closed without merging**